### PR TITLE
Slug redirections

### PR DESCRIPTION
**Pull request by noirbizarre**
This PR test, fix and enable slug redirections on any models with a permalink using the slug.

Slug (and their routing) where undertested (tested against `StringField` instead of `SlugField`) with false positive.

The following redirects will be enabled:
- UTF-8 slug redirects to the normalized one
- old slugs are redirected to the new one changes (until another object take the same slug)
- redirections are deleted on object deletion

### CODE REVIEW

File: udata/core/dataset/models.py [hunk 1]
```diff
@@ -360,8 +360,8 @@ class Dataset(WithMetrics, BadgeMixin, db.Owned, db.Document):
                                   default=datetime.now, required=True)
     title = db.StringField(required=True)
     acronym = db.StringField(max_length=128)
-    slug = db.SlugField(max_length=255, required=True,
-                        populate_from='title', update=True)
+    slug = db.SlugField(max_length=255, required=True, populate_from='title',
+                        update=True, follow=True)
```
Comment (abulte) [inline udata/core/dataset/models.py]: What does `follow` do?

File: udata/models/slug_fields.py [hunk 2]
```diff
@@ -83,19 +110,26 @@ def populate_slug(instance, field):
...
 
-    manual = (not previous and value or
-              field.db_field in instance._get_changed_fields())
+    # Field value has changed
+    changed = field.db_field in instance._get_changed_fields()
+    # Field initial value has been manually set
+    manual = not previous and value or changed
 
     if not manual and field.populate_from:
+        # value to slugify is extract from populate_from parameter
```
Comment (abulte) [inline udata/models/slug_fields.py]: s/extract/extracted/

File: udata/models/slug_fields.py [hunk 3]
```diff
@@ -130,14 +163,24 @@ def exists(s):
...
-
-        slug_follower.new_slug = slug
-        slug_follower.save()
+    if field.follow and old_slug != slug:
+        ns = instance.__class__.__name__
+        # Destroy redirections from this new slug
+        SlugFollow.objects(namespace=ns, old_slug=slug).delete()
+
+        if old_slug:
+            print('slugs', old_slug, slug)
```
Comment (abulte) [inline udata/models/slug_fields.py]: remove

File: udata/routing.py [hunk 4]
```diff
@@ -64,23 +71,47 @@ class ModelConverter(BaseConverter):
...
+
     def to_python(self, value):
         try:
-            obj = self.model.objects(slug=value).first()
-        except InvalidQueryError:  # If the model doesn't have a slug.
+            quoted = self.quote(value)
+            query = db.Q(slug=value) | db.Q(slug=quoted)
+            obj = self.model.objects(query).get()
+        except (InvalidQueryError, self.model.DoesNotExist):
+            # If the model doesn't have a slug or matching slug doesn't exists.
```
Comment (abulte) [inline udata/routing.py]: s/exists/exist/

File: udata/core/dataset/models.py (hunk 1 above)
Comment (abulte) [reply udata/core/dataset/models.py]: It's a custom trigger for the whole `SlugFollow` logic, right?

Comment (noirbizarre) [reply udata/core/dataset/models.py]: Yes `follow=True` enable redirections following slug changes. I kept the parameter name as it was

**Pull request merged**